### ISSUE TITLE
Third node is stuck in the loop after running the join command

### ISSUE THREAD

**Issue by jsemohub** [canonical/microk8s#3647]
#### Summary

Third node is stuck in the loop after running the join command
an  2 23:44:03 node2 microk8s.daemon-kubelite[356871]: W0102 23:44:03.416005  356871 logging.go:59] [core] [Channel #4 SubChannel #6] grpc: addrConn.createTransport failed to connect to {
Jan  2 23:44:03 node2 microk8s.daemon-kubelite[356871]:  "Addr": "unix:///var/snap/microk8s/4390/var/kubernetes/backend/kine.sock:12379",
Jan  2 23:44:03 node2 microk8s.daemon-kubelite[356871]:  "ServerName": "kine.sock",
Jan  2 23:44:03 node2 microk8s.daemon-kubelite[356871]:  "Attributes": null,
Jan  2 23:44:03 node2 microk8s.daemon-kubelite[356871]:  "BalancerAttributes": null,
Jan  2 23:44:03 node2 microk8s.daemon-kubelite[356871]:  "Type": 0,
Jan  2 23:44:03 node2 microk8s.daemon-kubelite[356871]:  "Metadata": null
Jan  2 23:44:03 node2 microk8s.daemon-kubelite[356871]: }. Err: connection error: desc = "transport: Error while dialing dial unix /var/snap/microk8s/4390/var/kubernetes/backend/kine.sock:12379: connect: no such file or directory"

#### Reproduction Steps

Yes. I can remove either of the nodes and the second join node always succeeds while the third always fails
1.  Run join command for node1. It succeeds.
2. Run join command for node2. It fails with the error message printed in the   

#### Introspection Report

Attached.
[inspection-report-20230102_234609.tar.gz](https://github.com/canonical/microk8s/files/10334367/inspection-report-20230102_234609.tar.gz)
#### Can you suggest a fix?

No.

**Comment by ktsakalozos** [comment]
Hi @jsemohub, could you offer some more info on your setup?

In particular it would be nice to have an inspection tarball of the node you are trying to join to so we can see how the rest of the network looks like. I would also be interested to know what the host OS is.

Looking at the logs, there are a few lines like:
```
Jan 02 21:32:26 node2.emopoll.com microk8s.daemon-kubelite[228771]: Trace[442139274]:  ---"Txn call succeeded" 17015ms (21:32:26.334)]
```
That must be before attempting to join, right? Taking 17 seconds to persist a db transaction is too long. It makes me wonder if there is some process hammering the disk or is the disk itself is not performing properly. If you have I/O intensive workloads you would better keep them away from the control plane nodes, this would mean you will need to join the worker nodes with the `--worker` flag in the `microk8s join` command.

Another source of unexpected behavior is when nodes do not resolve each other. Could you make sure all node hostnames are resolvable from all nodes. Also, the nodes should have a fixed IP that will never change. Finally make sure a node joins a cluster only once.

**Comment by jsemohub** [comment]
Thank you for the prompt response!

- When I switch the nodes, the problem occurs on another node. Harddrives are pretty fast, reliable (samsung pro) and stress tested. So they are unlikely to be the culprit. 
- Hostnames are resolvable. However, nodes are joinging on a private network and I am suspecting coreDNS might be highjacking BIND DNS traffic on slave servers.
- I tried `join` as a  worker mode. I had to  link to openssl.cnf  `openssl version -d` prints ssl home directory.  Perhaps, this could be used instead of hard coding openssl dir in python scripts. It seemed to work but requires further testing. Granted, high availability would be a more scalable option. 

From your comments, I am inclined to investigate further DNS related issues. 

Is coreDNS plugin used only for in-cluster name resolution or does it overwrite system's BIND config?

**Comment by jsemohub** [comment]
Some log snippets from  node1
`Jan  3 07:34:26 node1 microk8s.daemon-kubelite[793324]: Trace[1806813765]:  ---"Txn call completed" 11990ms (07:34:20.074)
Jan  3 07:34:26 node1 microk8s.daemon-kubelite[793324]: Trace[1806813765]:  ---"Txn call failed" err:context deadline exceeded 6736ms (07:34:26.811)]
Jan  3 07:34:26 node1 microk8s.daemon-kubelite[793324]: Trace[2118471708]:  ---"Txn call completed" 13178ms (07:34:26.906)]
Jan  3 07:34:26 node1 microk8s.daemon-kubelite[793324]: Trace[501896292]:  ---"Txn call completed" 13177ms (07:34:26.917)]
Jan  3 07:34:31 node1 microk8s.daemon-kubelite[793324]: Trace[1010583851]:  ---"Txn call completed" 14311ms (07:34:31.006)]
Jan  3 07:34:33 node1 microk8s.daemon-kubelite[793324]: Trace[741243768]:  ---"Txn call succeeded" 14690ms (07:34:33.050)]
Jan  3 07:34:33 node1 microk8s.daemon-kubelite[793324]: Trace[1807827353]:  ---"Txn call failed" err:context deadline exceeded 9998ms (07:34:33.418)]
Jan  3 07:34:33 node1 microk8s.daemon-kubelite[793324]: Trace[961516623]:  ---"Txn call failed" err:context deadline exceeded 9993ms (07:34:33.540)]
Jan  3 07:34:34 node1 microk8s.daemon-kubelite[793324]: Trace[1230084671]:  ---"Txn call completed" 13495ms (07:34:13.805)
Jan  3 07:34:34 node1 microk8s.daemon-kubelite[793324]: Trace[1230084671]:  ---"Txn call completed" 14832ms (07:34:28.638)
Jan  3 07:34:34 node1 microk8s.daemon-kubelite[793324]: Trace[1230084671]:  ---"Txn call failed" err:context deadline exceeded 5667ms (07:34:34.307)]
Jan  3 07:34:35 node1 microk8s.daemon-kubelite[793324]: Trace[523934596]:  ---"Txn call completed" 12962ms (07:34:21.800)
Jan  3 07:34:35 node1 microk8s.daemon-kubelite[793324]: Trace[523934596]:  ---"Txn call completed" 13300ms (07:34:35.102)]
Jan  3 07:34:37 node1 microk8s.daemon-kubelite[793324]: Trace[1025686182]:  ---"Txn call completed" 13886ms (07:34:37.146)]
Jan  3 07:34:41 node1 microk8s.daemon-kubelite[793324]: Trace[1040533529]:  ---"Txn call completed" 13959ms (07:34:40.889)
Jan  3 07:34:41 node1 microk8s.daemon-kubelite[793324]: Trace[1040533529]:  ---"Txn call failed" err:context canceled 1037ms (07:34:41.928)]
Jan  3 07:34:43 node1 microk8s.daemon-kubelite[793324]: Trace[1114312655]:  ---"Txn call failed" err:context deadline exceeded 9998ms (07:34:43.419)]
Jan  3 07:34:43 node1 microk8s.daemon-kubelite[793324]: Trace[18322578]:  ---"Txn call failed" err:context canceled 9993ms (07:34:43.558)]
Jan  3 07:34:45 node1 microk8s.daemon-kubelite[793324]: Trace[1068701205]:  ---"Txn call succeeded" 18430ms (07:34:45.341)]
Jan  3 07:34:46 node1 microk8s.daemon-kubelite[793324]: Trace[1716759062]:  ---"Txn call completed" 15331ms (07:34:46.363)]
Jan  3 07:34:48 node1 microk8s.daemon-kubelite[793324]: Trace[1513254021]:  ---"Txn call failed" err:rpc error: code = Unknown desc = exec (try: 500): database is locked 15013ms (07:34:48.067)]
Jan  3 07:34:51 node1 microk8s.daemon-kubelite[793324]: Trace[1649873399]: ---"Txn call completed" 17135ms (07:34:51.481)
`
Similar log snippets from node2
`Jan  2 21:52:39 node2 microk8s.daemon-kubelite[241464]: Trace[120865576]:  ---"Txn call completed" 11184ms (21:52:39.556)]
Jan  2 21:52:40 node2 microk8s.daemon-kubelite[241464]: Trace[1982993427]:  ---"Txn call completed" 11473ms (21:52:40.793)]
Jan  2 21:52:43 node2 microk8s.daemon-kubelite[241464]: Trace[540173426]:  ---"Txn call failed" err:rpc error: code = Unknown desc = exec (try: 500): database is locked 12794ms (21:52:43.927)]
Jan  2 21:52:45 node2 microk8s.daemon-kubelite[241464]: Trace[421680063]:  ---"Txn call failed" err:context deadline exceeded 9999ms (21:52:45.885)]
Jan  2 21:52:46 node2 microk8s.daemon-kubelite[241464]: Trace[1025395062]:  ---"Txn call failed" err:context deadline exceeded 9995ms (21:52:46.383)]
Jan  2 21:52:48 node2 microk8s.daemon-kubelite[241464]: Trace[1378723124]: ---"Txn call completed" 14105ms (21:52:48.983)
Jan  2 21:52:51 node2 microk8s.daemon-kubelite[241464]: Trace[1231452922]:  ---"Txn call completed" 10728ms (21:52:27.979)
Jan  2 21:52:51 node2 microk8s.daemon-kubelite[241464]: Trace[1231452922]:  ---"Txn call completed" 9993ms (21:52:37.979)
Jan  2 21:52:51 node2 microk8s.daemon-kubelite[241464]: Trace[1231452922]:  ---"Txn call completed" 12600ms (21:52:50.586)
Jan  2 21:52:51 node2 microk8s.daemon-kubelite[241464]: Trace[1231452922]:  ---"Txn call failed" err:context deadline exceeded 655ms (21:52:51.248)]
Jan  2 21:52:51 node2 microk8s.daemon-kubelite[241464]: Trace[398930308]:  ---"Txn call completed" 12180ms (21:52:29.548)
Jan  2 21:52:51 node2 microk8s.daemon-kubelite[241464]: Trace[398930308]:  ---"Txn call completed" 12824ms (21:52:42.373)
Jan  2 21:52:51 node2 microk8s.daemon-kubelite[241464]: Trace[398930308]:  ---"Txn call failed" err:context deadline exceeded 8993ms (21:52:51.367)]
Jan  2 21:52:52 node2 microk8s.daemon-kubelite[241464]: Trace[2080794705]:  ---"Txn call failed" err:rpc error: code = Unknown desc = exec (try: 500): database is locked 8989ms (21:52:52.919)]
Jan  2 21:52:55 node2 microk8s.daemon-kubelite[241464]: Trace[585783430]:  ---"Txn call completed" 10563ms (21:52:55.126)]
Jan  2 21:52:55 node2 microk8s.daemon-kubelite[241464]: Trace[271503525]:  ---"Txn call failed" err:context deadline exceeded 9998ms (21:52:55.885)]
Jan  2 23:15:32 node2 microk8s.daemon-kubelite[324119]: Trace[1254050788]:  ---"Txn call completed" 954ms (23:15:32.630)]
Jan  2 23:15:32 node2 microk8s.daemon-kubelite[324119]: Trace[1199854310]:  ---"Txn call completed" 691ms (23:15:32.720)]
`

Definitely something odd going on with dsqlite.

**Comment by ktsakalozos** [comment]
> Is coreDNS plugin used only for in-cluster name resolution or does it overwrite system's BIND config?

CoreDNS is used only by the in-cluster services.

**Comment by jsemohub** [comment]
I noticed that calico node didn't start. Unsure when exactly this started happening. Here is a small snippet from the log. 

2023-01-03 19:08:17.487 [ERROR][20] tunnel-ip-allocator/ipam_block_reader_writer.go 179: Failed to claim affinity error=resource does not exist: BlockAffinity(dev6.xxx.com-10-1-20-0-26) with error: the server could not find the requested resource (post BlockAffinities.crd.projectcalico.org) host="dev6.xxx.com" subnet=10.1.20.0/26
2023-01-03 19:08:17.488 [ERROR][20] tunnel-ip-allocator/ipam.go 558: Error claiming pending affinity error=resource does not exist: BlockAffinity(dev6.xxx.com-10-1-20-0-26) with error: the server could not find the requested resource (post BlockAffinities.crd.projectcalico.org) host="dev6.xxx.com" subnet=10.1.20.0/26
2023-01-03 19:08:17.488 [ERROR][20] tunnel-ip-allocator/ipam.go 119: Error assigning IPV4 addresses: resource does not exist: BlockAffinity(dev6.xxx.com-10-1-20-0-26) with error: the server could not find the requested resource (post BlockAffinities.crd.projectcalico.org)
2023-01-03 19:08:17.488 [FATAL][20] tunnel-ip-allocator/allocateip.go 472: Unable to autoassign an address error=resource does not exist: BlockAffinity(dev6.xxx.com-10-1-20-0-26) with error: the server could not find the requested resource (post BlockAffinities.crd.projectcalico.org) type="vxlanTunnelAddress"
Calico node failed to start

**Comment by jsemohub** [comment]
I was finally able to move past the node being stuck. Took a few runs and restarts. 
I on to the next one related to dns. It could have been related since I had to troubleshoot one addon at a time.

DNS fails to lookup ip
 k exec  -it nginx-ingress-microk8s-controller-m8ddh -n ingress -- bash
bash-5.1$ nslookup google.com
Server:         10.152.183.10
Address:        10.152.183.10:53

** server can't find google.com: SERVFAIL

It doesn't work.
Here looking at the resolve.
cat /etc/resolv.conf
search ingress.svc.cluster.local svc.cluster.local cluster.local
nameserver 10.152.183.10
options ndots:5

 k get ConfigMap coredns -n kube-system -o yaml
apiVersion: v1
data:
  Corefile: |
    .:53 {
        errors
        health {
          lameduck 5s
        }
        ready
        log . {
          class error
        }
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
        }
        prometheus :9153
        forward . 8.8.8.8 8.8.4.4
        cache 30
        loop
        reload
        loadbalance
    }
kind: ConfigMap
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","data":{"Corefile":".:53 {\n    errors\n    health {\n      lameduck 5s\n    }\n    ready\n    log . {\n      class error\n    }\n    kubernetes cluster.local in-addr.arpa ip6.arpa {\n      pods insecure\n      fallthrough in-addr.arpa ip6.arpa\n    }\n    prometheus :9153\n    forward . 8.8.8.8 8.8.4.4\n    cache 30\n    loop\n    reload\n    loadbalance\n}\n"},"kind":"ConfigMap","metadata":{"annotations":{},"labels":{"addonmanager.kubernetes.io/mode":"EnsureExists","k8s-app":"kube-dns"},"name":"coredns","namespace":"kube-system"}}
  creationTimestamp: "2023-01-05T00:41:09Z"
  labels:
    addonmanager.kubernetes.io/mode: EnsureExists
    k8s-app: kube-dns
  name: coredns
  namespace: kube-system
  resourceVersion: "259864"
  uid: 1b6fc72b-eabd-4139-b2de-9764d08b6553

I will file a separate ticket.
